flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricRegistry.java

```diff
@@ -209,14 +223,29 @@ public void shutdown() {
 			reporters = null;
 		}
 		shutdownExecutor();
+
+		if (stopFuture != null) {
```

Comment:
if the MQS is started while a shutdown is in progress it might not be cleaned up. What do you think about adding a lock here and in startMetricQueryService?